Versioning and Stability for OpenTelemetry (#1291)
open-telemetry/opentelemetry-specification: specification/overview.md

This makes OpenTelemetry a **cross-cutting concern** - a piece of software which is be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principle – separation of concerns. As a result, OpenTelemetry client design requires extra care and attention to avoid creating issues for the codebase which depend upon these cross-cutting APIs.

OpenTelemetry clients are designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions which can be managed independently. OpenTelemetry clients are also designed to be an extensible framework.
To accomplish this these goals, each signal consists of four types of packages: API, SDK, Semantic Conventions, and Contrib.

### API

API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of an OpenTelemetry client which is imported into 3rd-party libraries and application code is considered part of the API.

> To manage different levels of stability, every signal has its own, independent API package.

Reviewer: Is this mandatory? Or can all stable signals be in the main API package, and only experimental ones split out separately?

Author: Well, the issue is how to manage the transition from experimental to stable with minimal disruption. If packages must start out experimental, and thus in a separate package, then that separate package would probably stick around. 

For packages that exist as part of the initial v1.0, I'm not sure that the requirement matters, though it might be kind of ugly to have some packages to be laid out one way, and others another...

Reviewer 2: Recently thought about the experimental -> stable transition and am hoping that fair amount of disruption is fine, and is what a user signs up for when using the alpha versioned artifact. In fact, they will probably have disruption on minor versions since it's alpha - otherwise it would just be GA :-) For example of duplication, we have some duplication of API to metrics to support the version split - we have `GlobalOpenTelemetry` to manage a global `OpenTelemetry` which exposes all stable signals, and a `GlobalMeterProvider` to manage a global metrics signal because it can't be part of `OpenTelemetry` yet. We would like to remove `GlobalMeterProvider` when metrics goes GA - this will cause disruption though, and relatively significant disruption I think as it is basically changing the entry point to metrics.

I'm hoping that the only users of alpha signals are OpenTelemetry instrumentation itself, then we can manage disruption ourselves. And any potential end-user needs to know they're not going to have a good time - that's my expectation of the alpha signal and a versioned artifact like `1.3.1-alpha`.

Author: So, the problem is not alpha, but beta. The stability guarantee can only come after we have some production usage and feedback. Breaking alpha users is okay, but a huge break in support looming over the beta will discourage participation. Who could suggest adopting a beta in good conscience if the knew for a fact that all the code they wrote would break on them as part of the v1.0? I really would prefer there to be no breaking change between the final beta, the release candidate, and the v1.0 release, except for fixing the usage issues those releases are designed to catch.

Reviewer: I think the way we would do this is not to break anyone's code, but just remove the requirement to import that alpha package separately. Everything else (java packaging, naming, etc) would remain exactly the same when we move the code into the omnibus API module.

Reviewer 2: @jkwatson So are we committing to deprecating but maintaining until a future major version APIs we only added due to the artifact split, like `GlobalMeterProvider`? Ideally we didn't, but maybe we have to live with it.

Reviewer: right. we can deprecate things in the alpha module, and not carry them forward when we move everything into the main module.

Author: Ok, I understand. There are multiple approaches to achieving these guarantees, and keeping the experimental code from affecting stable code. Separate packages is one, the approach @jkwatson mentions is another. A third would be to keep experimental work on a branch - .NET would like to take this approach.

Ultimately, the goal is to meet the stability goals, not dictate a particular approach. I'll re-write this section to make the goals clearer, and emphasize that the separate packages is just one example approach.

Author: I've removed references to independent packages.

Application owners use the SDK constructors; [plugin authors](glossary.md#plugin-author) use the SDK plugin interfaces.
[Instrumentation authors](glossary.md#instrumentation-author) must never directly reference any SDK package of any kind, only the API.

### Semantic Conventions

The **Semantic Conventions** define the keys and values which describe commonly observed concepts, protocols, and operations used by applications.

* [Resource Conventions](resource/semantic_conventions/README.md)
* [Span Conventions](trace/semantic_conventions/README.md)
* [Metrics Conventions](metrics/semantic_conventions/README.md)

### Contrib Packages

The OpenTelemetry project maintains integrations with popular OSS projects which have been identified as important for observing modern web services.
Example API integrations include instrumentation for web frameworks, database clients, and message queues.
Example SDK integrations include plugins for exporting telemetry to popular analysis tools and telemetry storage systems.

Some plugins, such as OTLP Exporters and TraceContext Propagators, are required by the OpenTelemetry specification. These required plugins are included as part of the SDK.

> Plugins and instrumentation packages which are optional and separate from the SDK are referred to as **Contrib** packages.

Reviewer 3: Is this same as all the packages maintained in the -contrin github repos for individual languages?

Author: Correct. Everything we provide, but is not included by default.

</details>

This document provides an overview of the OpenTelemetry project and defines important fundamental terms.

Additional term definitions can be found in the [glossary](glossary.md).

> ## OpenTelemetry Client Architecture

Reviewer 4: @tedsuo Please also check if the wording introduced here matches what's defined in [library-guidelines.md](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md) and update it where needed (or consider removing duplicated sections).

Author: I made some improvements to this document to match the new information. At some point, I believe we could take another pass at [library-guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md) and [library-layout](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-layout.md). But I would like that to be a separate PR as it would greatly expand this PR, which is already rather large.

![Cross cutting concerns](../internal/img/architecture.png)

At the highest architectural level, OpenTelemetry clients are organized into **signals**. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three separate signals. Signals share a common subsystem – **context propagation** – but they function independently from each other.

> Each signal provides a mechanism for software to describe itself. A codebase, such as web framework or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principle – separation of concerns. As a result, OpenTelemetry client design requires extra care and attention to avoid creating issues for the codebase which depend upon these cross-cutting APIs.

Reviewer 4: > OpenTelemetry instrumentation code is then mixed into the other code within that codebase.

This is not the case for instrumentation libraries provided separately or even by OpenTelemetry SIGs themselves. How about:

```suggestion
Each signal provides a mechanism for software to describe itself. A codebase, such as web framework or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code can be mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which is mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principle – separation of concerns. As a result, OpenTelemetry client design requires extra care and attention to avoid creating issues for the codebase which depend upon these cross-cutting APIs.
```

PS: Adding line breaks after each sentence (or wherever semantically appropriate) would make commenting and tracking changes easier :)

Author: Thanks, will do. But do we have rules here? Seems like the goal of line breaks is to keep the width to roughly 80 characters, not to break things up semantically.

Reviewer 4: I remember we had some discussions around this topic but don't think there's a documented "style guide" or something like that. I think we agreed that we don't want to break sentences apart at some artificial hard limit but rather where it makes sense semantically to still get a nice diff (i.e., at the end of sentences and for long sentences between clauses). This makes changes easier to track (also in the blame view) and you  don't have 5 different threads/conversations for the same line if statements are split into separate lines.

Author: Sounds good. I formatted my portion in this manner, but left pre-existing content alone. This could be cleaned up in another PR.

OpenTelemetry clients are designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions which can be managed independently. OpenTelemetry clients are also designed to be an extensible framework. To accomplish this these goals, each signal consists of four types of packages.

### API

> API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of an OpenTelemetry client which imported into 3rd-party libraries and application code is considered part of the API. To manage different levels of stability, every signal has its own, independent API package. Stable APIs may then be bundled up into a unified API which provides additional convenience methods.

Reviewer 4: ```suggestion
API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of an OpenTelemetry client which is imported into 3rd-party libraries and application code is considered part of the API. To manage different levels of stability, every signal has its own, independent API package. Stable APIs may then be bundled up into a unified API which provides additional convenience methods.
```

### SDK

> The implementation of the API. The SDK is managed by the application owner. Note that the SDK includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces are defined as **constructors** and **plugin interfaces**. Examples of plugin interfaces include the `SpanProcessor`, `Exporter`, and `Sampler` interfaces. Examples of constructors include **configuration objects**, **environment variables**, and **SDK builders**. Application owners use the SDK constructors; plugin authors use the SDK plugin interfaces. Instrumentation authors must never directly reference any SDK package of any kind, only the API.

Reviewer 4: ```suggestion
The implementation of the API. The SDK is managed by the application owner. Note that the SDK includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces are defined as [**constructors**](glossary.md#constructors) and [**plugin interfaces**](glossary.md#sdk-plugins). Examples of plugin interfaces include the `SpanProcessor`, `Exporter`, and `Sampler` interfaces. Examples of constructors include **configuration objects**, **environment variables**, and **SDK builders**. Application owners use the SDK constructors; plugin authors use the SDK plugin interfaces. Instrumentation authors must never directly reference any SDK package of any kind, only the API.
```

### Semantics

> The **Semantic Conventions** define the keys and values which describe commonly observed concepts and operations within cloud computing. See details [below](#semantic-conventions).

Reviewer 4: ```suggestion
The **Semantic Conventions** define the keys and values which describe commonly observed concepts and operations in applications. See details [below](#semantic-conventions).
```
They also apply to software running in old-school (on-prem or remote) data centers and such.

### SDK

The implementation of the API. The SDK is managed by the application owner. Note that the SDKs includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces are defined as **constructors** and **plugin interfaces**. Examples of plugin interfaces include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders. Application owners may interact with SDK constructors; plugin authors may interact with SDK plugin interfaces. Instrumentation authors must never directly reference any SDK package of any kind, only the API.

### Semantics

**Semantic Conventions** are schema defining the attributes which describe common concepts and operations which the signal observes. See details [below](#semantic-conventions).

### Contrib Packages

> Plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel, are referred to as Contrib packages. The term "contrib" specifically refers to the plugins and instrumentation maintained by the OpenTelemetry organization outside of the SDK; it does not refer to third party plugins hosted elsewhere, or core plugins which are required to be part of the SDK release, such as OTLP Exporters and TraceContext Propagators. **API Contrib** refers to packages which depend solely upon the API; **SDK Contrib** refers to packages which also depend upon the SDK.

Reviewer 5: Running OTel => using OpenTelementry ? (No Otel definitely)
